lib/new_relic/agent/log_event_aggregator.rb

```diff
@@ -0,0 +1,185 @@
+# encoding: utf-8
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+
+require 'new_relic/agent/event_aggregator'
+require 'new_relic/agent/log_priority'
+
+module NewRelic
+  module Agent
+    class LogEventAggregator < EventAggregator
+
+      # Common keys
+      PLUGIN_TYPE_KEY = "plugin.type".freeze
+      PLUGIN_TYPE = "nr-ruby_agent".freeze
+
+      # Per-message keys
+      LEVEL_KEY = "level".freeze
+      MESSAGE_KEY = "message".freeze
+      TIMESTAMP_KEY = "timestamp".freeze
+
+      # Metric keys
+      LINES = "Logging/lines".freeze
+
+      named :LogEventAggregator
+      # TODO use the right value when the collector starts reporting it
+      capacity_key :'custom_insights_events.max_samples_stored'
+      #capacity_key :'log_sending.max_samples_stored'
+      enabled_key :'log_sending.enabled'
+      buffer_class PrioritySampledBuffer
+
+      def initialize(events)
+        super(events)
+        @counter_lock = Mutex.new
+        @seen = 0
+        @seen_by_severity = Hash.new(0)
+      end
+
+      def capacity
+        @buffer.capacity
+      end
+
+      def record(formatted_message, severity)
+        @counter_lock.synchronize do
+          @seen += 1
+          @seen_by_severity[severity] += 1
+        end
```

Comment:
These local values get tallied in the aggregator for recording as a metric each harvest. This avoids the expense of coercing metric names in every log call like we're doing in 8.1-8.4.